rbd/encryption.go

```diff
@@ -0,0 +1,116 @@
+// +build !octopus,!nautilus
+
+package rbd
+
+// #cgo LDFLAGS: -lrbd
+// /* force XSI-complaint strerror_r() */
+// #define _POSIX_C_SOURCE 200112L
+// #undef _GNU_SOURCE
+// #include <errno.h>
+// #include <stdlib.h>
+// #include <rados/librados.h>
+// #include <rbd/librbd.h>
+import "C"
+
+import (
+	"unsafe"
+)
+
+// EncryptionFormat: encryption format type
+type EncryptionFormat C.uint
+
+// Possible values for EncryptionFormat:
+// RBD_ENCRYPTION_FORMAT_LUKS1: LUKS v1
+// RBD_ENCRYPTION_FORMAT_LUKS2: LUKS v2
+const (
+	RBD_ENCRYPTION_FORMAT_LUKS1 = EncryptionFormat(C.RBD_ENCRYPTION_FORMAT_LUKS1)
+	RBD_ENCRYPTION_FORMAT_LUKS2 = EncryptionFormat(C.RBD_ENCRYPTION_FORMAT_LUKS2)
+)
+
+// EncryptionAlgorithm: encryption algorithm
+type EncryptionAlgorithm C.uint
+
+// Possible values for EncryptionAlgorithm:
+// RBD_ENCRYPTION_ALGORITHM_AES128: AES 128bits
+// RBD_ENCRYPTION_ALGORITHM_AES256: AES 256bits
+const (
+	RBD_ENCRYPTION_ALGORITHM_AES128 = EncryptionAlgorithm(C.RBD_ENCRYPTION_ALGORITHM_AES128)
+	RBD_ENCRYPTION_ALGORITHM_AES256 = EncryptionAlgorithm(C.RBD_ENCRYPTION_ALGORITHM_AES256)
+)
+
+// EncryptionFormatOptsLUKS1 and EncryptionFormatOptsLUKS2 are identical
+// structures at the moment, just as they are in the librbd api.
+// The purpose behind creating different identical structures, is to facilitate
+// future modifications of one of the formats, while maintaining backwards
+// compatibility with the other.
+
+// EncryptionFormatOptsLUKS1: Options required for LUKS v1
+type EncryptionFormatOptsLUKS1 struct {
+	Alg            EncryptionAlgorithm
+	Passphrase     string
+	PassphraseSize uint
```

Comment:
I am not sure you need the `PassphraseSize`, will it not be the same as `len(Passphrase)`?

Reply:
Come to think of it - do we know if the passphrases supported by ceph are valid unicode strings or does it support "random" binary data. If it's the latter the type should probably be `[]byte` rather than `string`. @effi-ofer is there any documentation you're aware of we can reference to determine this?

Reply:
Good point.  The LUKS official documentation allow for passing in the passphrase from a file, so I am *guessing* it can be binary data.  I am checking to confirm, but unless I update this issue assume it should be a []byte.